central/cve/telemetry/singleton.go

```diff
@@ -0,0 +1,92 @@
+package telemetry
+
+import (
+	"context"
+	"time"
+
+	deploymentDS "github.com/stackrox/rox/central/deployment/datastore"
+	"github.com/stackrox/rox/central/metrics"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var (
+	once     sync.Once
+	instance *vulnerabilityMetricsImpl
+)
+
+func Singleton() *vulnerabilityMetricsImpl {
+	once.Do(func() {
+		if env.AggregateVulnMetrics.Setting() == "" {
+			return
+		}
+		metricExpressions := parseAggregationExpressions(env.AggregateVulnMetrics.Setting())
+		if metricExpressions == nil {
+			return
+		}
+		instance = &vulnerabilityMetricsImpl{
+			ds:                deploymentDS.Singleton(),
+			metricExpressions: metricExpressions,
+			trackFunc:         metrics.SetAggregatedVulnCount,
+		}
+		for metricName, expressions := range instance.metricExpressions {
+			labels := getMetricLabels(expressions)
+			metrics.RegisterVulnAggregatedMetric(metricName, labels)
+		}
+	})
+	return instance
+}
+
+type metricName = string
+type metricKey = string // e.g. IMPORTANT_VULNERABILITY_SEVERITY|true
+
+type vulnerabilityMetricsImpl struct {
+	ds         deploymentDS.DataStore
+	stopSignal chan bool
+
+	// metricExpressions associates a metric name to the list of expressions.
+	//
+	// Example:
+	//
+	//   "Namespace_eq_abc_Severity_total": {"Namespace=abc", "Severity"},
+	metricExpressions map[metricName][]*expression
+	trackFunc         func(metricName string, labels map[Label]string, total int)
+}
+
+func (h *vulnerabilityMetricsImpl) Start() {
+	if h != nil {
+		go h.run()
+	}
+}
+
+func (h *vulnerabilityMetricsImpl) Stop() {
+	if h != nil {
+		close(h.stopSignal)
+	}
+}
```

Comment:
Calling Stop twice would cause panic. Maybe you could use `concurrency.Singnal` or `context.Context` instead.

Reply:
This is unfortunately the common central pattern. We do not call Stop or Start twice.

I am working on a better approach for handling periodic tasks, but that's a separate topic. I didn't want to introduce anything new in this PR.